```
Error running mint automation
+ mkdir -p 14177-730b5b2/gopath/src/github.com/minio
+ git -C 14177-730b5b2/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 14177-730b5b2/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 14177-730b5b2/gopath/src/github.com/minio/minio fetch --quiet minio pull/14177/head:pr14177
+ git -C 14177-730b5b2/gopath/src/github.com/minio/minio checkout --quiet pr14177
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/14177-730b5b2/gopath
+ make -C 14177-730b5b2/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/pkg v1.1.14
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/minio/minio-go/v7 v7.0.21
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/colinmarc/hdfs/v2 v2.2.0
go: downloading github.com/minio/cli v1.22.0
go: downloading google.golang.org/api v0.58.0
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985
go: downloading github.com/klauspost/compress v1.13.6
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading github.com/minio/kes v0.14.0
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/golang-jwt/jwt/v4 v4.1.0
go: downloading github.com/Shopify/sarama v1.27.2
go: downloading github.com/fatih/color v1.13.0
go: downloading golang.org/x/crypto v0.0.0-20211209193657-4570a0811e8b
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/klauspost/reedsolomon v1.9.15
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/minio/console v0.13.3
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/VividCortex/ewma v1.1.1
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/go-openapi/loads v0.21.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/gorilla/mux v1.8.0
go: downloading github.com/miekg/dns v1.1.43
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/minio/selfupdate v0.4.0
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.0+incompatible
go: downloading github.com/prometheus/client_golang v1.11.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/shirou/gopsutil/v3 v3.21.9
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.0
go: downloading go.etcd.io/etcd/api/v3 v3.5.0
go: downloading go.etcd.io/etcd/client/v3 v3.5.0
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20210819190943-2bc19b11175f
go: downloading go.uber.org/zap v1.19.1
go: downloading github.com/go-ldap/ldap/v3 v3.2.4
go: downloading github.com/coredns/coredns v1.4.0
go: downloading golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/eclipse/paho.mqtt.golang v1.3.5
go: downloading github.com/elastic/go-elasticsearch/v7 v7.12.0
go: downloading github.com/go-sql-driver/mysql v1.5.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.9.0
go: downloading github.com/nats-io/nats.go v1.11.1-0.20210623165838-4b75fc59ae30
go: downloading github.com/nats-io/stan.go v0.8.3
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/golang-jwt/jwt v3.2.2+incompatible
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/minio/simdjson-go v0.2.1
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
go: downloading github.com/mattn/go-colorable v0.1.10
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/pquerna/cachecontrol v0.0.0-20171018203845-0dec1b30a021
go: downloading gopkg.in/square/go-jose.v2 v2.3.1
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/klauspost/cpuid/v2 v2.0.9
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/minio/parquet-go v1.1.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading golang.org/x/net v0.0.0-20211112202133-69e39bad7dc2
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20200615164410-66371956d46c
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.1
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/go-openapi/analysis v0.20.1
go: downloading github.com/go-openapi/spec v0.20.4
go: downloading github.com/go-openapi/swag v0.19.15
go: downloading github.com/prometheus/common v0.31.1
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/lestrrat-go/jwx v1.2.7
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.0
go: downloading google.golang.org/grpc v1.41.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading google.golang.org/genproto v0.0.0-20210928142010-c7af6a1a74c9
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading golang.org/x/text v0.3.7
go: downloading github.com/gorilla/websocket v1.4.2
go: downloading github.com/golang/snappy v0.0.3
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/mailru/easyjson v0.7.6
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/tidwall/gjson v1.11.0
go: downloading github.com/tidwall/sjson v1.2.3
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading go.uber.org/multierr v1.7.0
go: downloading github.com/rs/xid v1.3.0
go: downloading gopkg.in/ini.v1 v1.63.2
go: downloading google.golang.org/protobuf v1.27.1
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading cloud.google.com/go v0.94.1
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading gopkg.in/jcmturner/gokrb5.v7 v7.5.0
go: downloading github.com/eapache/go-resiliency v1.2.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20200313005456-10cdbea86bc0
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading github.com/go-openapi/strfmt v0.21.1
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/jcmturner/goidentity/v6 v6.0.1
go: downloading github.com/mitchellh/mapstructure v1.4.1
go: downloading github.com/oklog/ulid v1.3.1
go: downloading go.mongodb.org/mongo-driver v1.7.5
go: downloading github.com/asaskevich/govalidator v0.0.0-20200907205600-7a23bdc65eef
go: downloading github.com/go-openapi/errors v0.20.1
go: downloading github.com/go-openapi/jsonreference v0.19.6
go: downloading github.com/PuerkitoBio/purell v1.1.1
go: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
go: downloading github.com/googleapis/gax-go/v2 v2.1.1
go: downloading go.opencensus.io v0.23.0
go: downloading github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
go: downloading github.com/go-stack/stack v1.8.0
go: downloading github.com/go-openapi/runtime v0.21.0
go: downloading github.com/jessevdk/go-flags v1.4.0
go: downloading github.com/minio/mc v0.0.0-20211207230606-23a05f5a17f2
go: downloading github.com/unrolled/secure v1.0.9
go: downloading github.com/minio/direct-csi v1.3.5-0.20210601185811-f7776f7961bf
go: downloading k8s.io/client-go v0.21.1
go: downloading github.com/minio/operator v0.0.0-20220110040724-a5d59a342b7f
go: downloading github.com/go-openapi/validate v0.20.3
go: downloading github.com/docker/go-units v0.4.0
go: downloading k8s.io/apimachinery v0.21.1
go: downloading k8s.io/kube-openapi v0.0.0-20210305001622-591a79e4bda7
go: downloading github.com/mb0/glob v0.0.0-20160210091149-1eb79d2de6c4
go: downloading github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: downloading github.com/briandowns/spinner v1.16.0
go: downloading github.com/pkg/xattr v0.4.3
go: downloading github.com/posener/complete v1.2.3
go: downloading maze.io/x/duration v0.0.0-20160924141736-faac084b6075
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/minio/colorjson v1.0.1
go: downloading github.com/pkg/profile v1.6.0
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/tklauser/go-sysconf v0.3.9
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/tklauser/numcpus v0.3.0
go: downloading sigs.k8s.io/controller-runtime v0.8.0
go: downloading k8s.io/api v0.21.1
go: downloading k8s.io/klog/v2 v2.8.0
go: downloading github.com/go-logr/logr v0.4.0
go: downloading sigs.k8s.io/yaml v1.2.0
go: downloading github.com/google/gofuzz v1.1.0
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.1.0
go: downloading github.com/google/go-cmp v0.5.6
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading github.com/emicklei/go-restful v2.9.5+incompatible
go: downloading gopkg.in/jcmturner/dnsutils.v1 v1.0.1
go: downloading gopkg.in/jcmturner/aescts.v1 v1.0.1
go: downloading gopkg.in/jcmturner/rpc.v1 v1.1.0
go: downloading k8s.io/utils v0.0.0-20201110183641-67b214c5f920
go: downloading github.com/googleapis/gnostic v0.5.1
go: downloading gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
../../../../pkg/mod/github.com/minio/operator@v0.0.0-20220110040724-a5d59a342b7f/pkg/apis/minio.min.io/v1/helper.go:49:2: github.com/minio/madmin-go@v1.2.7: replacement directory ../madmin-go does not exist
make: *** [build] Error 1

```

--- a/cmd/admin-handlers-site-replication.go
+++ b/cmd/admin-handlers-site-replication.go
@@ -338,8 +338,9 @@ func (a adminAPIHandlers) SiteReplicationStatus(w http.ResponseWriter, r *http.R
 	if objectAPI == nil {
 		return
 	}
+	opts := getSRStatusOptions(r)
 
-	info, err := globalSiteReplicationSys.SiteReplicationStatus(ctx, objectAPI)
+	info, err := globalSiteReplicationSys.SiteReplicationStatus(ctx, objectAPI, opts)
 	if err != nil {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
@@ -362,7 +363,8 @@ func (a adminAPIHandlers) SiteReplicationMetaInfo(w http.ResponseWriter, r *http
 		return
 	}
 
-	info, err := globalSiteReplicationSys.SiteReplicationMetaInfo(ctx, objectAPI)
+	opts := getSRStatusOptions(r)
+	info, err := globalSiteReplicationSys.SiteReplicationMetaInfo(ctx, objectAPI, opts)
 	if err != nil {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
@@ -427,4 +429,15 @@ func (a adminAPIHandlers) SRPeerEdit(w http.ResponseWriter, r *http.Request) {
 		writeErrorResponseJSON(ctx, w, toAdminAPIErr(ctx, err), r.URL)
 		return
 	}
+}
+
+func getSRStatusOptions(r *http.Request) (opts madmin.SRStatusOptions) {
+	q := r.Form
+	opts.Buckets = q.Get("buckets") == "true"
+	opts.Policies = q.Get("policies") == "true"
+	opts.Groups = q.Get("groups") == "true"
+	opts.Users = q.Get("users") == "true"
+	opts.Entity = madmin.GetSREntityType(q.Get("entity"))
+	opts.EntityValue = q.Get("entityvalue")
+	return
 }

--- a/cmd/site-replication.go
+++ b/cmd/site-replication.go
@@ -1930,8 +1930,18 @@ type srGroupPolicyMapping struct {
 	DeploymentID string
 }
 
+type srUserInfo struct {
+	madmin.UserInfo
+	DeploymentID string
+}
+
+type srGroupDesc struct {
+	madmin.GroupDesc
+	DeploymentID string
+}
+
 // SiteReplicationStatus returns the site replication status across clusters participating in site replication.
-func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI ObjectLayer) (info madmin.SRStatusInfo, err error) {
+func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI ObjectLayer, opts madmin.SRStatusOptions) (info madmin.SRStatusInfo, err error) {
 	c.RLock()
 	defer c.RUnlock()
 	if !c.enabled {
@@ -1949,7 +1959,7 @@ func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI O
 		index := index
 		if depIDs[index] == globalDeploymentID {
 			g.Go(func() error {
-				sris[index], sriErrs[index] = c.SiteReplicationMetaInfo(ctx, objAPI)
+				sris[index], sriErrs[index] = c.SiteReplicationMetaInfo(ctx, objAPI, opts)
 				return nil
 			}, index)
 			continue
@@ -1959,7 +1969,7 @@ func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI O
 			if err != nil {
 				return err
 			}
-			sris[index], sriErrs[index] = admClient.SRMetaInfo(ctx)
+			sris[index], sriErrs[index] = admClient.SRMetaInfo(ctx, opts)
 			return nil
 		}, index)
 	}
@@ -1990,6 +2000,8 @@ func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI O
 	policyStats := make(map[string][]srPolicy)
 	userPolicyStats := make(map[string][]srUserPolicyMapping)
 	groupPolicyStats := make(map[string][]srGroupPolicyMapping)
+	userInfoStats := make(map[string][]srUserInfo)
+	groupDescStats := make(map[string][]srGroupDesc)
 
 	numSites := len(sris)
 	for _, sri := range sris {
@@ -2012,234 +2024,395 @@ func (c *SiteReplicationSys) SiteReplicationStatus(ctx context.Context, objAPI O
 			userPolicyStats[user] = append(userPolicyStats[user], srUserPolicyMapping{SRPolicyMapping: policy, DeploymentID: sri.DeploymentID})
 		}
 		for group, policy := range sri.GroupPolicies {
-			if _, ok := userPolicyStats[group]; !ok {
+			if _, ok := groupPolicyStats[group]; !ok {
 				groupPolicyStats[group] = make([]srGroupPolicyMapping, 0, numSites)
 			}
 			groupPolicyStats[group] = append(groupPolicyStats[group], srGroupPolicyMapping{SRPolicyMapping: policy, DeploymentID: sri.DeploymentID})
 		}
+		for u, ui := range sri.UserInfoMap {
+			if _, ok := userInfoStats[u]; !ok {
+				userInfoStats[u] = make([]srUserInfo, 0, numSites)
+			}
+			userInfoStats[u] = append(userInfoStats[u], srUserInfo{UserInfo: ui, DeploymentID: sri.DeploymentID})
+		}
+		for g, gd := range sri.GroupDescMap {
+			if _, ok := groupDescStats[g]; !ok {
+				groupDescStats[g] = make([]srGroupDesc, 0, numSites)
+			}
+			groupDescStats[g] = append(groupDescStats[g], srGroupDesc{GroupDesc: gd, DeploymentID: sri.DeploymentID})
+		}
 	}
+
 	info.StatsSummary = make(map[string]madmin.SRSiteSummary, len(c.state.Peers))
-	info.BucketMismatches = make(map[string]map[string]madmin.SRBucketStatsSummary)
-	info.PolicyMismatches = make(map[string]map[string]madmin.SRPolicyStatsSummary)
-	info.UserMismatches = make(map[string]map[string]madmin.SRUserStatsSummary)
-	info.GroupMismatches = make(map[string]map[string]madmin.SRGroupStatsSummary)
+	info.BucketStats = make(map[string]map[string]madmin.SRBucketStatsSummary)
+	info.PolicyStats = make(map[string]map[string]madmin.SRPolicyStatsSummary)
+	info.UserStats = make(map[string]map[string]madmin.SRUserStatsSummary)
+	info.GroupStats = make(map[string]map[string]madmin.SRGroupStatsSummary)
 	// collect user policy mapping replication status across sites
-	for u, pslc := range userPolicyStats {
-		policySet := set.NewStringSet()
-		uPolicyCount := 0
-		for _, ps := range pslc {
-			policyBytes, err := json.Marshal(ps)
-			if err != nil {
-				continue
+	if opts.Users || opts.Entity == madmin.SRUserEntity {
+		for u, pslc := range userPolicyStats {
+			policySet := set.NewStringSet()
+			uPolicyCount := 0
+			for _, ps := range pslc {
+				policyBytes, err := json.Marshal(ps.SRPolicyMapping)
+				if err != nil {
+					continue
+				}
+				uPolicyCount++
+				sum := info.StatsSummary[ps.DeploymentID]
+				sum.TotalUserPolicyMappingCount++
+				info.StatsSummary[ps.DeploymentID] = sum
+
+				if policyStr := string(policyBytes); !policySet.Contains(policyStr) {
+					policySet.Add(policyStr)
+				}
 			}
-			uPolicyCount++
-			if policyStr := string(policyBytes); !policySet.Contains(policyStr) {
-				policySet.Add(policyStr)
+			userPolicyMismatch := !isReplicated(uPolicyCount, numSites, policySet)
+			switch {
+			case userPolicyMismatch, opts.Entity == madmin.SRUserEntity:
+				for _, ps := range pslc {
+					dID := depIdxes[ps.DeploymentID]
+					_, hasUser := sris[dID].UserPolicies[u]
+					if len(info.UserStats[u]) == 0 {
+						info.UserStats[u] = make(map[string]madmin.SRUserStatsSummary)
+					}
+					info.UserStats[u][ps.DeploymentID] = madmin.SRUserStatsSummary{
+						PolicyMismatch:   userPolicyMismatch,
+						HasUser:          hasUser,
+						HasPolicyMapping: ps.Policy != "",
+					}
+				}
+			default:
+				// no mismatch
+				for _, s := range pslc {
+					sum := info.StatsSummary[s.DeploymentID]
+					if !s.IsGroup {
+						sum.ReplicatedUserPolicyMappings++
+					}
+					info.StatsSummary[s.DeploymentID] = sum
+				}
 			}
 		}
-		policyMismatch := !isReplicated(uPolicyCount, numSites, policySet)
-		if policyMismatch {
+		// collect user info replication status across sites
+		for u, pslc := range userInfoStats {
+			uiSet := set.NewStringSet()
+			userCount := 0
 			for _, ps := range pslc {
-				dID := depIdxes[ps.DeploymentID]
-				_, hasUser := sris[dID].UserPolicies[u]
-
-				info.UserMismatches[u][ps.DeploymentID] = madmin.SRUserStatsSummary{
-					PolicyMismatch: policyMismatch,
-					UserMissing:    !hasUser,
+				uiBytes, err := json.Marshal(ps.UserInfo)
+				if err != nil {
+					continue
+				}
+				userCount++
+				sum := info.StatsSummary[ps.DeploymentID]
+				sum.TotalUsersCount++
+				info.StatsSummary[ps.DeploymentID] = sum
+				if uiStr := string(uiBytes); !uiSet.Contains(uiStr) {
+					uiSet.Add(uiStr)
+				}
+			}
+			userInfoMismatch := !isReplicated(userCount, numSites, uiSet)
+			switch {
+			case userInfoMismatch, opts.Entity == madmin.SRUserEntity:
+				for _, ps := range pslc {
+					dID := depIdxes[ps.DeploymentID]
+					_, hasUser := sris[dID].UserInfoMap[u]
+					if len(info.UserStats[u]) == 0 {
+						info.UserStats[u] = make(map[string]madmin.SRUserStatsSummary)
+					}
+					umis, ok := info.UserStats[u][ps.DeploymentID]
+					if !ok {
+						umis = madmin.SRUserStatsSummary{
+							HasUser: hasUser,
+						}
+					}
+					umis.UserInfoMismatch = userInfoMismatch
+					info.UserStats[u][ps.DeploymentID] = umis
+				}
+			default:
+				// no mismatch
+				for _, s := range pslc {
+					sum := info.StatsSummary[s.DeploymentID]
+					sum.ReplicatedUsers++
+					info.StatsSummary[s.DeploymentID] = sum
 				}
 			}
 		}
 	}
-
-	// collect user policy mapping replication status across sites
-
-	for g, pslc := range groupPolicyStats {
-		policySet := set.NewStringSet()
-		gPolicyCount := 0
-		for _, ps := range pslc {
-			policyBytes, err := json.Marshal(ps)
-			if err != nil {
-				continue
+	if opts.Groups || opts.Entity == madmin.SRGroupEntity {
+		// collect group policy mapping replication status across sites
+		for g, pslc := range groupPolicyStats {
+			policySet := set.NewStringSet()
+			gPolicyCount := 0
+			for _, ps := range pslc {
+				policyBytes, err := json.Marshal(ps.SRPolicyMapping)
+				if err != nil {
+					continue
+				}
+				gPolicyCount++
+				if policyStr := string(policyBytes); !policySet.Contains(policyStr) {
+					policySet.Add(policyStr)
+				}
+				sum := info.StatsSummary[ps.DeploymentID]
+				sum.TotalGroupPolicyMappingCount++
+				info.StatsSummary[ps.DeploymentID] = sum
 			}
-			gPolicyCount++
-			if policyStr := string(policyBytes); !policySet.Contains(policyStr) {
-				policySet.Add(policyStr)
+			groupPolicyMismatch := !isReplicated(gPolicyCount, numSites, policySet)
+
+			switch {
+			case groupPolicyMismatch, opts.Entity == madmin.SRGroupEntity:
+				for _, ps := range pslc {
+					dID := depIdxes[ps.DeploymentID]
+					_, hasGroup := sris[dID].GroupPolicies[g]
+					if len(info.GroupStats[g]) == 0 {
+						info.GroupStats[g] = make(map[string]madmin.SRGroupStatsSummary)
+					}
+					info.GroupStats[g][ps.DeploymentID] = madmin.SRGroupStatsSummary{
+						PolicyMismatch:   groupPolicyMismatch,
+						HasGroup:         hasGroup,
+						HasPolicyMapping: ps.Policy != "",
+					}
+				}
+			default:
+				// no mismatch
+				for _, s := range pslc {
+					sum := info.StatsSummary[s.DeploymentID]
+					sum.ReplicatedGroupPolicyMappings++
+					info.StatsSummary[s.DeploymentID] = sum
+				}
 			}
 		}
-		policyMismatch := !isReplicated(gPolicyCount, numSites, policySet)
-		if policyMismatch {
-			for _, ps := range pslc {
-				dID := depIdxes[ps.DeploymentID]
-				_, hasGroup := sris[dID].GroupPolicies[g]
 
-				info.GroupMismatches[g][ps.DeploymentID] = madmin.SRGroupStatsSummary{
-					PolicyMismatch: policyMismatch,
-					GroupMissing:   !hasGroup,
+		// collect group desc replication status across sites
+		for g, pslc := range groupDescStats {
+			gdSet := set.NewStringSet()
+			groupCount := 0
+			for _, ps := range pslc {
+				gdBytes, err := json.Marshal(ps.GroupDesc)
+				if err != nil {
+					continue
+				}
+				groupCount++
+				sum := info.StatsSummary[ps.DeploymentID]
+				sum.TotalGroupsCount++
+				info.StatsSummary[ps.DeploymentID] = sum
+
+				if gdStr := string(gdBytes); !gdSet.Contains(gdStr) {
+					gdSet.Add(gdStr)
+				}
+			}
+			gdMismatch := !isReplicated(groupCount, numSites, gdSet)
+			switch {
+			case gdMismatch, opts.Entity == madmin.SRGroupEntity:
+				for _, ps := range pslc {
+					dID := depIdxes[ps.DeploymentID]
+					_, hasGroup := sris[dID].GroupDescMap[g]
+					if len(info.GroupStats[g]) == 0 {
+						info.GroupStats[g] = make(map[string]madmin.SRGroupStatsSummary)
+					}
+					gmis, ok := info.GroupStats[g][ps.DeploymentID]
+					if !ok {
+						gmis = madmin.SRGroupStatsSummary{
+							HasGroup: hasGroup,
+						}
+					} else {
+						gmis.GroupDescMismatch = gdMismatch
+					}
+					info.GroupStats[g][ps.DeploymentID] = gmis
+				}
+			default:
+				// no mismatch
+				for _, s := range pslc {
+					sum := info.StatsSummary[s.DeploymentID]
+					sum.ReplicatedGroups++
+					info.StatsSummary[s.DeploymentID] = sum
 				}
 			}
 		}
 	}
-	// collect IAM policy replication status across sites
-
-	for p, pslc := range policyStats {
-		var policies []*iampolicy.Policy
-		uPolicyCount := 0
-		for _, ps := range pslc {
-			plcy, err := iampolicy.ParseConfig(bytes.NewReader(ps.policy))
-			if err != nil {
-				continue
-			}
-			policies = append(policies, plcy)
-			uPolicyCount++
-			sum := info.StatsSummary[ps.DeploymentID]
-			sum.TotalIAMPoliciesCount++
-			info.StatsSummary[ps.DeploymentID] = sum
-		}
-		policyMismatch := !isIAMPolicyReplicated(uPolicyCount, numSites, policies)
-		switch {
-		case policyMismatch:
+	if opts.Policies || opts.Entity == madmin.SRPolicyEntity {
+		// collect IAM policy replication status across sites
+		for p, pslc := range policyStats {
+			var policies []*iampolicy.Policy
+			uPolicyCount := 0
 			for _, ps := range pslc {
-				dID := depIdxes[ps.DeploymentID]
-				_, hasPolicy := sris[dID].Policies[p]
-				if len(info.PolicyMismatches[p]) == 0 {
-					info.PolicyMismatches[p] = make(map[string]madmin.SRPolicyStatsSummary)
-				}
-				info.PolicyMismatches[p][ps.DeploymentID] = madmin.SRPolicyStatsSummary{
-					PolicyMismatch: policyMismatch,
-					PolicyMissing:  !hasPolicy,
-				}
-			}
-		default:
-			// no mismatch
-			for _, s := range pslc {
-				sum := info.StatsSummary[s.DeploymentID]
-				if !policyMismatch {
-					sum.ReplicatedIAMPolicies++
-				}
-				info.StatsSummary[s.DeploymentID] = sum
-			}
-
-		}
-	}
-	// collect bucket metadata replication stats across sites
-	for b, slc := range bucketStats {
-		tagSet := set.NewStringSet()
-		olockConfigSet := set.NewStringSet()
-		var policies []*bktpolicy.Policy
-		var replCfgs []*sreplication.Config
-		sseCfgSet := set.NewStringSet()
-		var tagCount, olockCfgCount, sseCfgCount int
-		for _, s := range slc {
-			if s.ReplicationConfig != nil {
-				cfgBytes, err := base64.StdEncoding.DecodeString(*s.ReplicationConfig)
-				if err != nil {
-					continue
-				}
-				cfg, err := sreplication.ParseConfig(bytes.NewReader(cfgBytes))
-				if err != nil {
-					continue
-				}
-				replCfgs = append(replCfgs, cfg)
-			}
-			if s.Tags != nil {
-				tagBytes, err := base64.StdEncoding.DecodeString(*s.Tags)
-				if err != nil {
-					continue
-				}
-				tagCount++
-				if !tagSet.Contains(string(tagBytes)) {
-					tagSet.Add(string(tagBytes))
-				}
-			}
-			if len(s.Policy) > 0 {
-				plcy, err := bktpolicy.ParseConfig(bytes.NewReader(s.Policy), b)
+				plcy, err := iampolicy.ParseConfig(bytes.NewReader(ps.policy))
 				if err != nil {
 					continue
 				}
 				policies = append(policies, plcy)
+				uPolicyCount++
+				sum := info.StatsSummary[ps.DeploymentID]
+				sum.TotalIAMPoliciesCount++
+				info.StatsSummary[ps.DeploymentID] = sum
 			}
-			if s.ObjectLockConfig != nil {
-				olockCfgCount++
-				if !olockConfigSet.Contains(*s.ObjectLockConfig) {
-					olockConfigSet.Add(*s.ObjectLockConfig)
+			policyMismatch := !isIAMPolicyReplicated(uPolicyCount, numSites, policies)
+			switch {
+			case policyMismatch, opts.Entity == madmin.SRPolicyEntity:
+				for _, ps := range pslc {
+					dID := depIdxes[ps.DeploymentID]
+					_, hasPolicy := sris[dID].Policies[p]
+					if len(info.PolicyStats[p]) == 0 {
+						info.PolicyStats[p] = make(map[string]madmin.SRPolicyStatsSummary)
+					}
+					info.PolicyStats[p][ps.DeploymentID] = madmin.SRPolicyStatsSummary{
+						PolicyMismatch: policyMismatch,
+						HasPolicy:      hasPolicy,
+					}
 				}
-			}
-			if s.SSEConfig != nil {
-				if !sseCfgSet.Contains(*s.SSEConfig) {
-					sseCfgSet.Add(*s.SSEConfig)
+			default:
+				// no mismatch
+				for _, s := range pslc {
+					sum := info.StatsSummary[s.DeploymentID]
+					if !policyMismatch {
+						sum.ReplicatedIAMPolicies++
+					}
+					info.StatsSummary[s.DeploymentID] = sum
 				}
-				sseCfgCount++
+
 			}
-			ss, ok := info.StatsSummary[s.DeploymentID]
-			if !ok {
-				ss = madmin.SRSiteSummary{}
-			}
-			// increment total number of replicated buckets
-			if len(slc) == numSites {
-				ss.ReplicatedBuckets++
-			}
-			ss.TotalBucketsCount++
-			if tagCount > 0 {
-				ss.TotalTagsCount++
-			}
-			if olockCfgCount > 0 {
-				ss.TotalLockConfigCount++
-			}
-			if sseCfgCount > 0 {
-				ss.TotalSSEConfigCount++
-			}
-			if len(policies) > 0 {
-				ss.TotalBucketPoliciesCount++
-			}
-			info.StatsSummary[s.DeploymentID] = ss
 		}
-		tagMismatch := !isReplicated(tagCount, numSites, tagSet)
-		olockCfgMismatch := !isReplicated(olockCfgCount, numSites, olockConfigSet)
-		sseCfgMismatch := !isReplicated(sseCfgCount, numSites, sseCfgSet)
-		policyMismatch := !isBktPolicyReplicated(numSites, policies)
-		replCfgMismatch := !isBktReplCfgReplicated(numSites, replCfgs)
-		switch {
-		case tagMismatch, olockCfgMismatch, sseCfgMismatch, policyMismatch, replCfgMismatch:
-			info.BucketMismatches[b] = make(map[string]madmin.SRBucketStatsSummary, numSites)
+	}
+	if opts.Buckets || opts.Entity == madmin.SRBucketEntity {
+		// collect bucket metadata replication stats across sites
+		for b, slc := range bucketStats {
+			tagSet := set.NewStringSet()
+			olockConfigSet := set.NewStringSet()
+			var policies []*bktpolicy.Policy
+			var replCfgs []*sreplication.Config
+			var quotaCfgs []*madmin.BucketQuota
+			sseCfgSet := set.NewStringSet()
+			var tagCount, olockCfgCount, sseCfgCount int
 			for _, s := range slc {
-				dID := depIdxes[s.DeploymentID]
-				_, hasBucket := sris[dID].Buckets[s.Bucket]
-				info.BucketMismatches[b][s.DeploymentID] = madmin.SRBucketStatsSummary{
-					DeploymentID:           s.DeploymentID,
-					HasBucket:              hasBucket,
-					TagMismatch:            tagMismatch,
-					OLockConfigMismatch:    olockCfgMismatch,
-					SSEConfigMismatch:      sseCfgMismatch,
-					PolicyMismatch:         policyMismatch,
-					ReplicationCfgMismatch: replCfgMismatch,
-					HasReplicationCfg:      len(replCfgs) > 0,
+				if s.ReplicationConfig != nil {
+					cfgBytes, err := base64.StdEncoding.DecodeString(*s.ReplicationConfig)
+					if err != nil {
+						continue
+					}
+					cfg, err := sreplication.ParseConfig(bytes.NewReader(cfgBytes))
+					if err != nil {
+						continue
+					}
+					replCfgs = append(replCfgs, cfg)
 				}
+				if s.QuotaConfig != nil {
+					cfgBytes, err := base64.StdEncoding.DecodeString(*s.QuotaConfig)
+					if err != nil {
+						continue
+					}
+					cfg, err := parseBucketQuota(b, cfgBytes)
+					if err != nil {
+						continue
+					}
+					quotaCfgs = append(quotaCfgs, cfg)
+				}
+				if s.Tags != nil {
+					tagBytes, err := base64.StdEncoding.DecodeString(*s.Tags)
+					if err != nil {
+						continue
+					}
+					tagCount++
+					if !tagSet.Contains(string(tagBytes)) {
+						tagSet.Add(string(tagBytes))
+					}
+				}
+				if len(s.Policy) > 0 {
+					plcy, err := bktpolicy.ParseConfig(bytes.NewReader(s.Policy), b)
+					if err != nil {
+						continue
+					}
+					policies = append(policies, plcy)
+				}
+				if s.ObjectLockConfig != nil {
+					olockCfgCount++
+					if !olockConfigSet.Contains(*s.ObjectLockConfig) {
+						olockConfigSet.Add(*s.ObjectLockConfig)
+					}
+				}
+				if s.SSEConfig != nil {
+					if !sseCfgSet.Contains(*s.SSEConfig) {
+						sseCfgSet.Add(*s.SSEConfig)
+					}
+					sseCfgCount++
+				}
+				ss, ok := info.StatsSummary[s.DeploymentID]
+				if !ok {
+					ss = madmin.SRSiteSummary{}
+				}
+				// increment total number of replicated buckets
+				if len(slc) == numSites {
+					ss.ReplicatedBuckets++
+				}
+				ss.TotalBucketsCount++
+				if tagCount > 0 {
+					ss.TotalTagsCount++
+				}
+				if olockCfgCount > 0 {
+					ss.TotalLockConfigCount++
+				}
+				if sseCfgCount > 0 {
+					ss.TotalSSEConfigCount++
+				}
+				if len(policies) > 0 {
+					ss.TotalBucketPoliciesCount++
+				}
+				info.StatsSummary[s.DeploymentID] = ss
 			}
-			fallthrough
-		default:
-			// no mismatch
-			for _, s := range slc {
-				sum := info.StatsSummary[s.DeploymentID]
-				if !olockCfgMismatch && olockCfgCount == numSites {
-					sum.ReplicatedLockConfig++
+			tagMismatch := !isReplicated(tagCount, numSites, tagSet)
+			olockCfgMismatch := !isReplicated(olockCfgCount, numSites, olockConfigSet)
+			sseCfgMismatch := !isReplicated(sseCfgCount, numSites, sseCfgSet)
+			policyMismatch := !isBktPolicyReplicated(numSites, policies)
+			replCfgMismatch := !isBktReplCfgReplicated(numSites, replCfgs)
+			quotaCfgMismatch := !isBktQuotaCfgReplicated(numSites, quotaCfgs)
+			switch {
+			case tagMismatch, olockCfgMismatch, sseCfgMismatch, policyMismatch, replCfgMismatch, quotaCfgMismatch, opts.Entity == madmin.SRBucketEntity:
+				info.BucketStats[b] = make(map[string]madmin.SRBucketStatsSummary, numSites)
+				for i, s := range slc {
+					dID := depIdxes[s.DeploymentID]
+					_, hasBucket := sris[dID].Buckets[s.Bucket]
+
+					info.BucketStats[b][s.DeploymentID] = madmin.SRBucketStatsSummary{
+						DeploymentID:           s.DeploymentID,
+						HasBucket:              hasBucket,
+						TagMismatch:            tagMismatch,
+						OLockConfigMismatch:    olockCfgMismatch,
+						SSEConfigMismatch:      sseCfgMismatch,
+						PolicyMismatch:         policyMismatch,
+						ReplicationCfgMismatch: replCfgMismatch,
+						QuotaCfgMismatch:       quotaCfgMismatch,
+						HasReplicationCfg:      s.ReplicationConfig != nil,
+						HasTagsSet:             s.Tags != nil,
+						HasOLockConfigSet:      s.ObjectLockConfig != nil,
+						HasPolicySet:           s.Policy != nil,
+						HasQuotaCfgSet:         *quotaCfgs[i] != madmin.BucketQuota{},
+						HasSSECfgSet:           s.SSEConfig != nil,
+					}
 				}
-				if !sseCfgMismatch && sseCfgCount == numSites {
-					sum.ReplicatedSSEConfig++
+				fallthrough
+			default:
+				// no mismatch
+				for _, s := range slc {
+					sum := info.StatsSummary[s.DeploymentID]
+					if !olockCfgMismatch && olockCfgCount == numSites {
+						sum.ReplicatedLockConfig++
+					}
+					if !sseCfgMismatch && sseCfgCount == numSites {
+						sum.ReplicatedSSEConfig++
+					}
+					if !policyMismatch && len(policies) == numSites {
+						sum.ReplicatedBucketPolicies++
+					}
+					if !tagMismatch && tagCount == numSites {
+						sum.ReplicatedTags++
+					}
+					info.StatsSummary[s.DeploymentID] = sum
 				}
-				if !policyMismatch && len(policies) == numSites {
-					sum.ReplicatedBucketPolicies++
-				}
-				if !tagMismatch && tagCount == numSites {
-					sum.ReplicatedTags++
-				}
-				info.StatsSummary[s.DeploymentID] = sum
 			}
 		}
 	}
 	// maximum buckets users etc seen across sites
 	info.MaxBuckets = len(bucketStats)
-	info.MaxUsers = len(userPolicyStats)
-	info.MaxGroups = len(groupPolicyStats)
+	info.MaxUsers = len(userInfoStats)
+	info.MaxGroups = len(groupDescStats)
 	info.MaxPolicies = len(policyStats)
 	return
 }
@@ -2271,6 +2444,26 @@ func isIAMPolicyReplicated(cntReplicated, total int, policies []*iampolicy.Polic
 			continue
 		}
 		if !prev.Equals(*p) {
+			return false
+		}
+	}
+	return true
+}
+
+func isBktQuotaCfgReplicated(total int, quotaCfgs []*madmin.BucketQuota) bool {
+	if len(quotaCfgs) > 0 && len(quotaCfgs) != total {
+		return false
+	}
+	var prev *madmin.BucketQuota
+	for i, q := range quotaCfgs {
+		if q == nil {
+			return false
+		}
+		if i == 0 {
+			prev = q
+			continue
+		}
+		if prev.Quota != q.Quota || prev.Type != q.Type {
 			return false
 		}
 	}
@@ -2333,116 +2526,153 @@ func isBktReplCfgReplicated(total int, cfgs []*sreplication.Config) bool {
 }
 
 // SiteReplicationMetaInfo returns the metadata info on buckets, policies etc for the replicated site
-func (c *SiteReplicationSys) SiteReplicationMetaInfo(ctx context.Context, objAPI ObjectLayer) (info madmin.SRInfo, err error) {
+func (c *SiteReplicationSys) SiteReplicationMetaInfo(ctx context.Context, objAPI ObjectLayer, opts madmin.SRStatusOptions) (info madmin.SRInfo, err error) {
 	if objAPI == nil {
 		return info, errSRObjectLayerNotReady
 	}
-
 	c.RLock()
 	defer c.RUnlock()
 	if !c.enabled {
 		return info, nil
 	}
-	buckets, err := objAPI.ListBuckets(ctx)
-	if err != nil {
-		return info, errSRBackendIssue(err)
-	}
 	info.DeploymentID = globalDeploymentID
+	if opts.Buckets || opts.Entity == madmin.SRBucketEntity {
+		var (
+			buckets []BucketInfo
+			err     error
+		)
+		if opts.Entity == madmin.SRBucketEntity {
+			bi, err := objAPI.GetBucketInfo(ctx, opts.EntityValue)
+			if err != nil {
+				return info, nil
+			}
+			buckets = append(buckets, bi)
+		} else {
+			buckets, err = objAPI.ListBuckets(ctx)
+			if err != nil {
+				return info, errSRBackendIssue(err)
+			}
+		}
+		info.Buckets = make(map[string]madmin.SRBucketInfo, len(buckets))
+		for _, bucketInfo := range buckets {
+			bucket := bucketInfo.Name
+			bms := madmin.SRBucketInfo{Bucket: bucket}
+			// Get bucket policy if present.
+			policy, err := globalPolicySys.Get(bucket)
+			found := true
+			if _, ok := err.(BucketPolicyNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			if found {
+				policyJSON, err := json.Marshal(policy)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				bms.Policy = policyJSON
+			}
 
-	info.Buckets = make(map[string]madmin.SRBucketInfo, len(buckets))
-	for _, bucketInfo := range buckets {
-		bucket := bucketInfo.Name
-		bms := madmin.SRBucketInfo{Bucket: bucket}
-		// Get bucket policy if present.
-		policy, err := globalPolicySys.Get(bucket)
-		found := true
-		if _, ok := err.(BucketPolicyNotFound); ok {
-			found = false
-		} else if err != nil {
-			return info, errSRBackendIssue(err)
-		}
-		if found {
-			policyJSON, err := json.Marshal(policy)
-			if err != nil {
-				return info, wrapSRErr(err)
+			// Get bucket tags if present.
+			tags, err := globalBucketMetadataSys.GetTaggingConfig(bucket)
+			found = true
+			if _, ok := err.(BucketTaggingNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
 			}
-			bms.Policy = policyJSON
-		}
+			if found {
+				tagBytes, err := xml.Marshal(tags)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				tagCfgStr := base64.StdEncoding.EncodeToString(tagBytes)
+				bms.Tags = &tagCfgStr
+			}
 
-		// Get bucket tags if present.
-		tags, err := globalBucketMetadataSys.GetTaggingConfig(bucket)
-		found = true
-		if _, ok := err.(BucketTaggingNotFound); ok {
-			found = false
-		} else if err != nil {
-			return info, errSRBackendIssue(err)
-		}
-		if found {
-			tagBytes, err := xml.Marshal(tags)
-			if err != nil {
-				return info, wrapSRErr(err)
+			// Get object-lock config if present.
+			objLockCfg, err := globalBucketMetadataSys.GetObjectLockConfig(bucket)
+			found = true
+			if _, ok := err.(BucketObjectLockConfigNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
 			}
-			tagCfgStr := base64.StdEncoding.EncodeToString(tagBytes)
-			bms.Tags = &tagCfgStr
-		}
+			if found {
+				objLockCfgData, err := xml.Marshal(objLockCfg)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				objLockStr := base64.StdEncoding.EncodeToString(objLockCfgData)
+				bms.ObjectLockConfig = &objLockStr
+			}
 
-		// Get object-lock config if present.
-		objLockCfg, err := globalBucketMetadataSys.GetObjectLockConfig(bucket)
-		found = true
-		if _, ok := err.(BucketObjectLockConfigNotFound); ok {
-			found = false
-		} else if err != nil {
-			return info, errSRBackendIssue(err)
-		}
-		if found {
-			objLockCfgData, err := xml.Marshal(objLockCfg)
-			if err != nil {
-				return info, wrapSRErr(err)
+			// Get quota config if present
+			quotaConfig, err := globalBucketMetadataSys.GetQuotaConfig(bucket)
+			found = true
+			if _, ok := err.(BucketQuotaConfigNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
 			}
-			objLockStr := base64.StdEncoding.EncodeToString(objLockCfgData)
-			bms.ObjectLockConfig = &objLockStr
-		}
+			if found {
+				quotaConfigJSON, err := json.Marshal(quotaConfig)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				quotaConfigStr := base64.StdEncoding.EncodeToString(quotaConfigJSON)
+				bms.QuotaConfig = &quotaConfigStr
+			}
 
-		// Get existing bucket bucket encryption settings
-		sseConfig, err := globalBucketMetadataSys.GetSSEConfig(bucket)
-		found = true
-		if _, ok := err.(BucketSSEConfigNotFound); ok {
-			found = false
-		} else if err != nil {
-			return info, errSRBackendIssue(err)
-		}
-		if found {
-			sseConfigData, err := xml.Marshal(sseConfig)
-			if err != nil {
-				return info, wrapSRErr(err)
+			// Get existing bucket bucket encryption settings
+			sseConfig, err := globalBucketMetadataSys.GetSSEConfig(bucket)
+			found = true
+			if _, ok := err.(BucketSSEConfigNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
 			}
-			sseConfigStr := base64.StdEncoding.EncodeToString(sseConfigData)
-			bms.SSEConfig = &sseConfigStr
-		}
-		// Get replication config if present
-		rcfg, err := globalBucketMetadataSys.GetReplicationConfig(ctx, bucket)
-		found = true
-		if _, ok := err.(BucketReplicationConfigNotFound); ok {
-			found = false
-		} else if err != nil {
-			return info, errSRBackendIssue(err)
-		}
-		if found {
-			rcfgXML, err := xml.Marshal(rcfg)
-			if err != nil {
-				return info, wrapSRErr(err)
+			if found {
+				sseConfigData, err := xml.Marshal(sseConfig)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				sseConfigStr := base64.StdEncoding.EncodeToString(sseConfigData)
+				bms.SSEConfig = &sseConfigStr
 			}
-			rcfgXMLStr := base64.StdEncoding.EncodeToString(rcfgXML)
-			bms.ReplicationConfig = &rcfgXMLStr
+
+			// Get replication config if present
+			rcfg, err := globalBucketMetadataSys.GetReplicationConfig(ctx, bucket)
+			found = true
+			if _, ok := err.(BucketReplicationConfigNotFound); ok {
+				found = false
+			} else if err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			if found {
+				rcfgXML, err := xml.Marshal(rcfg)
+				if err != nil {
+					return info, wrapSRErr(err)
+				}
+				rcfgXMLStr := base64.StdEncoding.EncodeToString(rcfgXML)
+				bms.ReplicationConfig = &rcfgXMLStr
+			}
+			info.Buckets[bucket] = bms
 		}
-		info.Buckets[bucket] = bms
 	}
 
-	{
-		// Replicate IAM policies on local to all peers.
-		allPolicies, err := globalIAMSys.ListPolicies(ctx, "")
-		if err != nil {
-			return info, errSRBackendIssue(err)
+	if opts.Policies || opts.Entity == madmin.SRPolicyEntity {
+		var allPolicies map[string]iampolicy.Policy
+		if opts.Entity == madmin.SRPolicyEntity {
+			if p, err := globalIAMSys.store.GetPolicy(opts.EntityValue); err == nil {
+				allPolicies = map[string]iampolicy.Policy{opts.EntityValue: p}
+			}
+		} else {
+			// Replicate IAM policies on local to all peers.
+			allPolicies, err = globalIAMSys.ListPolicies(ctx, "")
+			if err != nil {
+				return info, errSRBackendIssue(err)
+			}
 		}
 		info.Policies = make(map[string]json.RawMessage, len(allPolicies))
 		for pname, policy := range allPolicies {
@@ -2454,22 +2684,25 @@ func (c *SiteReplicationSys) SiteReplicationMetaInfo(ctx context.Context, objAPI
 		}
 	}
 
-	{
+	if opts.Users || opts.Entity == madmin.SRUserEntity {
 		// Replicate policy mappings on local to all peers.
 		userPolicyMap := make(map[string]MappedPolicy)
-		groupPolicyMap := make(map[string]MappedPolicy)
-		globalIAMSys.store.rlock()
-		errU := globalIAMSys.store.loadMappedPolicies(ctx, stsUser, false, userPolicyMap)
-		errG := globalIAMSys.store.loadMappedPolicies(ctx, stsUser, true, groupPolicyMap)
-		globalIAMSys.store.runlock()
-		if errU != nil {
-			return info, errSRBackendIssue(errU)
+		if opts.Entity == madmin.SRUserEntity {
+			if mp, ok := globalIAMSys.store.GetMappedPolicy(opts.EntityValue, false); ok {
+				userPolicyMap[opts.EntityValue] = mp
+			}
+		} else {
+			globalIAMSys.store.rlock()
+			if err := globalIAMSys.store.loadMappedPolicies(ctx, stsUser, false, userPolicyMap); err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			if err = globalIAMSys.store.loadMappedPolicies(ctx, regUser, false, userPolicyMap); err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			globalIAMSys.store.runlock()
 		}
-		if errG != nil {
-			return info, errSRBackendIssue(errG)
-		}
+
 		info.UserPolicies = make(map[string]madmin.SRPolicyMapping, len(userPolicyMap))
-		info.GroupPolicies = make(map[string]madmin.SRPolicyMapping, len(c.state.Peers))
 		for user, mp := range userPolicyMap {
 			info.UserPolicies[user] = madmin.SRPolicyMapping{
 				IsGroup:     false,
@@ -2477,12 +2710,68 @@ func (c *SiteReplicationSys) SiteReplicationMetaInfo(ctx context.Context, objAPI
 				Policy:      mp.Policies,
 			}
 		}
+		info.UserInfoMap = make(map[string]madmin.UserInfo)
+		if opts.Entity == madmin.SRUserEntity {
+			if ui, err := globalIAMSys.GetUserInfo(ctx, opts.EntityValue); err == nil {
+				info.UserInfoMap[opts.EntityValue] = ui
+			}
+		} else {
+			//  get users/group info on local.
+			userInfoMap, errU := globalIAMSys.ListUsers()
+			if errU != nil {
+				return info, errSRBackendIssue(errU)
+			}
+			for user, ui := range userInfoMap {
+				info.UserInfoMap[user] = ui
+			}
+		}
+	}
+	if opts.Groups || opts.Entity == madmin.SRGroupEntity {
+		// Replicate policy mappings on local to all peers.
+		groupPolicyMap := make(map[string]MappedPolicy)
+		if opts.Entity == madmin.SRUserEntity {
+			if mp, ok := globalIAMSys.store.GetMappedPolicy(opts.EntityValue, true); ok {
+				groupPolicyMap[opts.EntityValue] = mp
+			}
+		} else {
+			globalIAMSys.store.rlock()
+			if err := globalIAMSys.store.loadMappedPolicies(ctx, stsUser, true, groupPolicyMap); err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			if err := globalIAMSys.store.loadMappedPolicies(ctx, regUser, true, groupPolicyMap); err != nil {
+				return info, errSRBackendIssue(err)
+			}
+			globalIAMSys.store.runlock()
+		}
 
+		info.GroupPolicies = make(map[string]madmin.SRPolicyMapping, len(c.state.Peers))
 		for group, mp := range groupPolicyMap {
-			info.UserPolicies[group] = madmin.SRPolicyMapping{
+			info.GroupPolicies[group] = madmin.SRPolicyMapping{
 				IsGroup:     true,
 				UserOrGroup: group,
 				Policy:      mp.Policies,
+			}
+		}
+		info.GroupDescMap = make(map[string]madmin.GroupDesc)
+		if opts.Entity == madmin.SRGroupEntity {
+			if gd, err := globalIAMSys.GetGroupDescription(opts.EntityValue); err == nil {
+				info.GroupDescMap[opts.EntityValue] = gd
+			}
+		} else {
+			//  get users/group info on local.
+			groups, errG := globalIAMSys.ListGroups(ctx)
+			if errG != nil {
+				return info, errSRBackendIssue(errG)
+			}
+			groupDescMap := make(map[string]madmin.GroupDesc, len(groups))
+			for _, g := range groups {
+				groupDescMap[g], errG = globalIAMSys.GetGroupDescription(g)
+				if errG != nil {
+					return info, errSRBackendIssue(errG)
+				}
+			}
+			for group, d := range groupDescMap {
+				info.GroupDescMap[group] = d
 			}
 		}
 	}

--- a/go.mod
+++ b/go.mod
@@ -48,7 +48,7 @@ require (
 	github.com/minio/csvparser v1.0.0
 	github.com/minio/highwayhash v1.0.2
 	github.com/minio/kes v0.14.0
-	github.com/minio/madmin-go v1.2.8
+	github.com/minio/madmin-go v1.2.9
 	github.com/minio/minio-go/v7 v7.0.21
 	github.com/minio/parquet-go v1.1.0
 	github.com/minio/pkg v1.1.15
@@ -84,9 +84,9 @@ require (
 	go.etcd.io/etcd/client/v3 v3.5.0
 	go.uber.org/atomic v1.9.0
 	go.uber.org/zap v1.19.1
-	golang.org/x/crypto v0.0.0-20220112180741-5e0467b6c7ce
+	golang.org/x/crypto v0.0.0-20220128200615-198e4374d7ed
 	golang.org/x/oauth2 v0.0.0-20210819190943-2bc19b11175f
-	golang.org/x/sys v0.0.0-20220114195835-da31bd327af9
+	golang.org/x/sys v0.0.0-20220128215802-99c3d69c2c27
 	golang.org/x/time v0.0.0-20210723032227-1f47c861a9ac
 	google.golang.org/api v0.58.0
 	gopkg.in/yaml.v2 v2.4.0
@@ -200,7 +200,7 @@ require (
 	go.mongodb.org/mongo-driver v1.7.5 // indirect
 	go.opencensus.io v0.23.0 // indirect
 	go.uber.org/multierr v1.7.0 // indirect
-	golang.org/x/net v0.0.0-20220121210141-e204ce36a2ba // indirect
+	golang.org/x/net v0.0.0-20220127200216-cd36cc0744dd // indirect
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c // indirect
 	golang.org/x/term v0.0.0-20210927222741-03fcf44c2211 // indirect
 	golang.org/x/text v0.3.7 // indirect

